Product Creation AI: Generate product info

selanthiraiyan opened the pull request:
Closes: #10773 
Closes: #10771  

## Description

Identify language, generate product details and set the generated product into the view model. 

#### Changes

- Make shipping values optional and skip asking AI about it if dimension and weight units are unavailable.
- Set `Product` model’s `regularPrice` property instead of `price`.
- Set a default AITone if not already set. 
- Finetune AI prompt's price generation part. Internal - p1695773759702609-slack-C03L1NF1EA3
- Identify language and generate product.
- Add/update unit tests.


## Testing instructions

- Create a JN site with the Jetpack AI plugin. (Checkbox available on JN home page to install this plugin)
- Login into the app using the JN site
- Switch to the Products tab and select the "+" button to add a new product. 
- Tap "Create a product with AI" option
- Enter a product name and tap "Continue"
- Enter product features in the next screen and tap "Create Product Details"
- Optionally tap on "Set tone and voice" and select a tone.
- The product details should be displayed on the next screen after the loading animation.

## Screenshots


https://github.com/woocommerce/woocommerce-ios/assets/524475/fa74346d-872c-4143-a692-8d54d1e56ec8



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Store .casual as default AI tone.

selanthiraiyan committed:
Remove redundant code handle `nil` tone.

selanthiraiyan committed:
Make dimension and weight unit optional.

selanthiraiyan committed:
As for shipping info only if dimension and weight unit available.

selanthiraiyan committed:
Add empty Shipping value as decode fallback.

selanthiraiyan committed:
Add mock file to reflect no shipping info response.

selanthiraiyan committed:
Test that response with shipping object is parsed correctly.

selanthiraiyan committed:
Inject dependencies to read categories, tags and AI tone.

selanthiraiyan committed:
Helper methods to identify language and generate product.

selanthiraiyan committed:
Identify language and generate product.

selanthiraiyan committed:
Unit test generate details in ProductDetailPreviewViewModel

selanthiraiyan committed:
Save generated price to regularPrice property.

selanthiraiyan committed:
Update price generation part of AI prompt.

selanthiraiyan committed:
Merge branch 'trunk' into feat/10773-generate-product

itsmeichigo commented on Networking/Networking/Remote/GenerativeContentRemote.swift:
```diff
@@ -223,6 +223,20 @@ private extension GenerativeContentRemote {
         }()
 
+        let shippingJson = {
+            guard let dimensionUnit,
+                  let weightUnit else {
+                return ""
+            }
```
❓ Do we really need to require both dimension and weight units? Is there any case where either of them are available? E.g., when only the weight unit is available, we can still ask AI to guess the weight right?

selanthiraiyan replied:
Good thinking. Started asking based on available unit in e2f1054b01ffa508dfa5e4ad487afbf5c0bfcbc9 and added tests in d34e69194a30115e6b4184cb5fb13984e9901343 and d34e69194a30115e6b4184cb5fb13984e9901343

itsmeichigo commented on Networking/Networking/Remote/GenerativeContentRemote.swift:
````diff
@@ -238,14 +252,8 @@ private extension GenerativeContentRemote {
-                "}, " +
-            "price: Guess the price in ```\(currencySymbol)```, return the price unformatted" +
+            shippingJson +
+            "price: Guess the price in ```\(currencySymbol)```, do not include the currency symbol, only provide the price as a number" +
````
Super nit: Missing a comma `,` at the end of the line.

selanthiraiyan replied:
This `,` for JSON formatting will be added at the next line by the property `tagsAsString` when it is not empty. https://github.com/woocommerce/woocommerce-ios/pull/10781/files#diff-f3c1e0fc4a88c0ba5380eb28042e350607ba267796f2ebfebdc2c551501a702aR270

selanthiraiyan committed:
Make generatedProduct private(set) for unit tests.

selanthiraiyan committed:
Add `onProductCreated` callback in unit tests.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -170,10 +198,57 @@ private extension ProductDetailPreviewViewModel {
+        try await withCheckedThrowingContinuation { continuation in
+            stores.dispatch(ProductAction.generateProduct(siteID: siteID,
+                                                          productName: productName,
+                                                          keywords: productFeatures ?? "",
```
❓ Do we need to include also the product description here if it's available?

selanthiraiyan replied:
Great catch. Fixed in 94d5e9b2b1e8b1b1d61baf11c64f6dce932e5579

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -209,3 +284,7 @@ private extension ProductDetailPreviewViewModel {
     }
 }
+
+private enum IdentifyLanguageError: Error {
```
Nit: Should we keep this error internal instead of duplicating it in different features?

selanthiraiyan replied:
I agree. Done in 94d5e9b2b1e8b1b1d61baf11c64f6dce932e5579

itsmeichigo reviewed (approved):
This works great!

selanthiraiyan committed:
Ask for weight and dimensions based on each unit availability.

selanthiraiyan committed:
Mock files to test JSON response without weight/dimensions info.

selanthiraiyan committed:
Unit test that JSON without dimensions/weight info is parsed correctly.

selanthiraiyan committed:
If available send product description to generate product.

selanthiraiyan committed:
Reuse `IdentifyLanguageError` from Product description feature.

selanthiraiyan commented:
Thank you for the review, @itsmeichigo! 🙇 

I have addressed your comments and will enable auto-merge now to unblock development work. Please feel free to add any comments which I will address in future PRs.
